packages/fxa-auth-server/lib/payments/customer-locations.ts

```diff
@@ -0,0 +1,173 @@
+    // 1. Get the Stripe customer object from Firestore else Stripe
+    const customer = await this.stripeHelper.fetchCustomer(uid);
+    if (!customer) {
+      this.log.debug('customerLocations.getPayPalCustomerCountry', {
+        message: 'No customer found in Stripe for the user.',
+        uid,
+      });
+      return null;
+    }
+    // 2. Get the Stripe customer billing address; should be `address.country`
+    // We already know this user is or was a PayPal subscriber, so we
+    // don't have to check.
+    const { address } = customer;
+    const country = address?.country;
+    if (!country) {
+      this.log.debug('customerLocations.getPayPalCustomerCountry', {
+        message: 'No country found on customer address.',
+        customerId: customer.id,
+      });
+      return null;
+    }
+    return country;
+  }
+
+  /**
+   * Transform country long name into country short name.
+   */
+  mapLongToShortCountryName(country: string): string {
+    if (Object.keys(COUNTRIES_LONG_NAME_TO_SHORT_NAME_MAP).includes(country)) {
+      return COUNTRIES_LONG_NAME_TO_SHORT_NAME_MAP[country];
+    }
+    // We don't care about any other countries.
+    this.log.debug('customerLocations.mapLongToShortCountryName', {
+      message: 'Country not found in long name to short name map.',
+      country,
+    });
+    return country;
+  }
+
+  /**
+   * Transform state long name into state short name.
+   */
+  mapLongToShortStateName(country: string, state: string): string {
+    if (
+      Object.keys(STATES_LONG_NAME_TO_SHORT_NAME_MAP).includes(country) &&
+      Object.keys(STATES_LONG_NAME_TO_SHORT_NAME_MAP[country]).includes(state)
+    ) {
+      return STATES_LONG_NAME_TO_SHORT_NAME_MAP[country][state];
+    }
+    // We don't care about any other country/state combos.
+    this.log.debug('customerLocations.mapLongToShortStateName', {
+      message: 'State not found in long name to short name map.',
+      state,
+    });
+    return state;
+  }
+
+  /**
+   * Finds the best `country` and `state` location for the user based on a provided
+   * list of inferred locations and the country in their existing Stripe customer
+   * record.
+   *  - If no country is found in Stripe, `state` and `country` are both "unknown".
+   *  - If there is a country mismatch, prefer the country in Stripe and select the most
+   *    frequent location with a matching country.
+   *  - If there is no matching country, `state` is "unknown".
+   */
+  async getBestLocationForPayPalUser(
+    uid: string,
+    results: PayPalUserLocationResult[]
+  ): Promise<PayPalUserLocationResult | null> {
+    let bestLocation = { uid, state: 'unknown', country: 'unknown', count: 1 };
+    // Avoid hitting Stripe 100 rps rate limit in prod by artificially slowing down requests.
+    // After the first runthrough of the script, we should have backfilled everyone in
+    // Firestore, and this delay can be removed.
+    await sleep(200);
```

Comment:
Why is this in here?  This is a function for one user that makes one request iiuc.

Reply:
Yeah I left this in from when I did the [spike ticket](https://mozilla-hub.atlassian.net/browse/FXA-4218), in which I iterated through 100 uids at a time for 40,000+ uids. I can remove this so long as we remember to add it or something similar back in #11070 .